File: 백준/Silver/1780. 종이의 개수/종이의 개수.py
import sys
input = sys.stdin.readline

n = int(input())


def get_cnt(x,y,n):
  global color1, color2, color3
  div_index = n//3
  tmp = graph[x][y]
  for i in range(x,x+n):
    for j in range(y,y+n):
      if(tmp != graph[i][j]):
        get_cnt(x,y,div_index)
        get_cnt(x+div_index, y, div_index)
        get_cnt(x+2*div_index, y, div_index)
        get_cnt(x, y+div_index, div_index)
        get_cnt(x, y+2*div_index, div_index)
        get_cnt(x+div_index, y+div_index, div_index)
        get_cnt(x+div_index, y+2*div_index, div_index)
        get_cnt(x+2*div_index, y+div_index, div_index)
        get_cnt(x+2*div_index, y+2*div_index,div_index)
        return

  if(tmp == -1):
    color1 += 1

  elif(tmp == 0):
    color2 += 1

  else:
    color3 += 1
  
color1 = 0
color2 = 0 
color3 = 0
graph = []
for _ in range(n):
  graph.append(list(map(int, input().strip().split())))

get_cnt(0,0,n)
print(color1)
print(color2)
print(color3)

File: 백준/Silver/1780. 종이의 개수/종이의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static int color1;
    static int color2;
    static int color3;
    static int[][] data;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        data = new int[n][n];

        for(int i = 0; i<n; i++){
            String[] strings = br.readLine().split(" ");
            for(int j = 0; j<n; j++){
                data[i][j] = Integer.parseInt(strings[j]);
            }
        }
        cntColor(0,0,n);
        System.out.println(color1);
        System.out.println(color2);
        System.out.println(color3);
        br.close();
    }

    public static void cntColor(int a, int b, int n){
        int divIndex = n/3;

        int tmp = data[a][b];
        for(int i = a; i< (a+n); i++) {
            for (int j = b; j < (b + n); j++) {
                if (tmp != data[i][j]) {
                    for (int x = 0; x < 3; x++) {
                        for (int y = 0; y < 3; y++) {
                            cntColor(a + x * divIndex, b + y * divIndex, divIndex);
                        }
                    }
                    return;
                }
            }
        }
        if(tmp == -1){
            color1++;
        } else if(tmp == 0){
            color2++;
        } else if(tmp == 1){
            color3++;
        }
    }
}
